modules/renter/downloadbuffer.go

```diff
@@ -0,0 +1,172 @@
+// the stream.
+var errFailedStreamWrite = errors.New("downloadDestinationWriter has a broken stream due to a prior failed write")
+
+// newDownloadDestinationWriter takes a writer and converts it into a
+func newDownloadDestinationWriter(w io.WriteCloser) downloadDestination {
+	return &downloadDestinationWriter{WriteCloser: w}
+}
+
+// nextWrite will iterate over all of the blocking write calls and unblock the
+// one that is next in line, if the next-in-line call is available.
+func (ddw *downloadDestinationWriter) nextWrite() {
+	for i, offset := range ddw.blockingWriteCalls {
+		if offset == ddw.progress {
+			ddw.blockingWriteSignals[i].Unlock()
+			ddw.blockingWriteCalls = append(ddw.blockingWriteCalls[0:i], ddw.blockingWriteCalls[i+1:]...)
+			ddw.blockingWriteSignals = append(ddw.blockingWriteSignals[0:i], ddw.blockingWriteSignals[i+1:]...)
+			return
+		}
+		if offset < ddw.progress {
+			// Sanity check - there should not be a call to WriteAt that occurs
+			// earlier than the current progress. If there is, the
+			// downloadDestinationWriter is being used incorrectly in an
+			// unrecoverable way.
+			panic("incorrect write order for downloadDestinationWriter")
+		}
+	}
+}
+
+// WriteAt will block until the stream has progressed to or past 'offset', and
+// then it will write its own data.
+func (ddw *downloadDestinationWriter) WriteAt(data []byte, offset int64) (int, error) {
+	write := func() (int, error) {
+		// If the stream writer has already failed, return an error.
+		if ddw.failed {
+			return 0, errFailedStreamWrite
+		}
+
+		// Write the data to the stream.
+		n, err := ddw.Write(data)
+		if err != nil {
+			// If there is an error, marked the stream write as failed and then
+			// unlock/unblock all of the waiting WriteAt calls.
+			ddw.failed = true
+			ddw.Close()
+			for i := range ddw.blockingWriteSignals {
+				ddw.blockingWriteSignals[i].Unlock()
+			}
+			return n, err
+		}
+
+		// Update the progress and unblock the next write.
+		ddw.progress += int64(n)
+		ddw.nextWrite()
+		return n, nil
+	}
+
+	ddw.mu.Lock()
+	// Check if the stream has already failed. If so, return immediately with
+	// the failed stream error.
+	if ddw.failed {
+		ddw.mu.Unlock()
+		return 0, errFailedStreamWrite
+	}
+
+	// Check if we are writing to the correct offset for the stream. If so, call
+	// write() and return.
+	if offset == ddw.progress {
+		// This write is the next write in line.
+		n, err := write()
+		ddw.mu.Unlock()
+		return n, err
+	}
+
+	// Block until we are the correct offset for the stream. The blocking is
+	// coordinated by a new mutex which gets added to an array. When the earlier
+	// data is written, the mutex will be unlocked, allowing us to write.
+	myMu := new(sync.Mutex)
+	myMu.Lock()
+	ddw.blockingWriteCalls = append(ddw.blockingWriteCalls, offset)
+	ddw.blockingWriteSignals = append(ddw.blockingWriteSignals, myMu)
+	ddw.mu.Unlock()
+	myMu.Lock()
```

Comment:
Where is myMu.Unlock()?

Reply:
It's unlocked in a different thread that clears the slices and writes the data.

Reply:
docstring for this function should probably explain how that works, @ChrisSchinnerl was very baffled the first time he saw it as well